osu.Server.BeatmapSubmission/Services/S3BeatmapStorage.cs

```diff
@@ -0,0 +1,138 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using Amazon;
+using Amazon.Runtime;
+using Amazon.S3;
+using Amazon.S3.Model;
+using osu.Framework.Extensions;
+using osu.Game.IO.Archives;
+using osu.Server.BeatmapSubmission.Configuration;
+using osu.Server.BeatmapSubmission.Models;
+using SharpCompress.Writers;
+using SharpCompress.Writers.Zip;
+
+namespace osu.Server.BeatmapSubmission.Services
+{
+    public class S3BeatmapStorage : IBeatmapStorage
+    {
+        private const string osz_directory = "osz";
+        private const string versioned_file_directory = "beatmap_files";
```

Comment:
Please double-check the directory structure of the target S3 bucket here, and that this matches it. I only have limited visibility over it as web-10 source does not seem to contain the entirety of interactions with S3. I only surmised the existence of the `osz` directory from how mirrors operate, and web-10 upload flow doesn't actually seem to reference that directory. I suspect it's obscured by that other service that web-10 calls into to convert osz2 to osz.

Reply:
The `osz` part is correct, yes.

As for the new `beatmap_files`, it's maybe fine to exist like this for now? We can move it in the future as required. I'll probably want to do a check on the s3 structure before we end up deploying this though – we may decide we want this new storage to be on R2 or DO spaces instead, and it may make sense for it to be in an isolated bucket to allow local rules to be applied to it (storage redundancy / what not).

One other thing to note is that the main `.osu` files also need to be uploaded to a different s3 bucket (i think it's just called `beatmaps` or something, i'll look into this later if you don't already have visibility, let me know)

Reply:
> One other thing to note is that the main `.osu` files also need to be uploaded to a different s3 bucket (i think it's just called `beatmaps` or something, i'll look into this later if you don't already have visibility, let me know)

Oh yeah that is there in web-10 source, I must have glossed over it / forgot it... I'll get that part.

Reply:
https://github.com/ppy/osu-server-beatmap-submission/pull/13/commits/7097b6c591f9455be621db60e852961c3ab018b8 should do it I believe?

Reply:
The unfortunate part is I think the `beatmaps` folder is in a different bucket name. It's currently root folder in a bucket named `beatmaps`. I dunno how worthwhile it is having this in configuration, but probably where it should live for now.

![2024-12-25 19 46 08@2x](https://github.com/user-attachments/assets/be9756a3-4c6e-4431-9c35-0bd6a299cae4)

![2024-12-25 19 46 24@2x](https://github.com/user-attachments/assets/8c052015-3abf-4344-8b83-053598cae45e)

Reply:
Right, I appear to have completely misread your original comment. Will fix...

> I dunno how worthwhile it is having this in configuration

I've been trying to adhere to the principle of having everything external in envvars, since in the past I've been annoyed with some of our projects hardcoding references to external services and thus for instance accidentally hitting production LIO endpoint because I forgot to modify a thing. So I'm kinda following web's precedent of moving all of that out to a single place where everything external is configured.

Reply:
fd39d35fa6a4fdfbefb0009c92778867302fe869 is hopefully closer to what the desired result is :sweat_smile:

Reply:
Yep, this looks correct now.